.github/workflows/haskell-ci.yml

```diff
@@ -0,0 +1,208 @@
+        required: false
+        default: true
+      compat:
+        description: Whether to run the Cabal/GHC compatibility check.
+        type: boolean
+        required: false
+        default: true
+      compat-dirs:
+        description: Directories in which to run the Cabal/GHC compatibility check, relative to the repo root.
+        type: string
+        required: false
+        default: "."
+      configure-flags:
+        description: Flags to pass to `cabal configure`, e.g., `--ghc-options='-Werror=default -j'`
+        type: string
+        required: false
+        # For an explanation of `--ghc-options='-j'`, see Note [Parallelism].
+        default: --ghc-options='-j' --enable-tests
+      ghc:
+        description: GHC version, e.g., "9.10.1"
+        type: string
+        required: false
+        default: latest
+      haddock:
+        description: Whether to run `cabal haddock`.
+        type: boolean
+        required: false
+        default: true
+      os:
+        description: |
+          Runner OS, e.g., "ubuntu-latest".
+
+          Set this to the value of `runs-on`, *not* `runner.os`.
+        type: string
+        required: true
+      sdist:
+        description: |
+          Whether to run `cabal sdist`.
+
+          This will also cause later steps to be run in the unpacked sdist.
+        type: boolean
+        required: false
+        default: true
+      test:
+        description: Whether to run `cabal test`.
+        type: boolean
+        required: false
+        default: true
+      test-targets:
+        description: Cabal targets to test, e.g. `test:my-tests`.
+        type: string
+        required: false
+        default: all
+
+# General commentary
+# ------------------
+#
+# Steps that are less likely to fail should come later, because failing earlier
+# saves time and resources. For example, `cabal check` should come after `cabal
+# build`, because edits to the Cabal file are generally more rare than edits to
+# Haskell source files.
+jobs:
+  CI:
+    runs-on: ${{ inputs.os }}
+    steps:
+    - name: Checkout
+      uses: actions/checkout@v4
+    - name: Install Cabal and GHC
+      uses: haskell-actions/setup@v2
+      id: setup-haskell
+      with:
+        cabal-version: ${{ inputs.cabal }}
+        ghc-version: ${{ inputs.ghc }}
+    - name: Create sdist
+      if: inputs.sdist
+      run: |
+        if [[ "${{ inputs.sdist }}" == true ]]; then
+          cabal sdist -o sdist
+          mkdir -p unpacked
+          tar -C unpacked -xzf sdist/*.tar.gz
+          cd unpacked/*
+        fi
+        echo "WORK_DIR=$(pwd)" >> $GITHUB_ENV
+    - name: Restore Cabal cache
+      uses: actions/cache/restore@v4
+      id: cache
+      env:
+        # Each `inputs.os` (e.g., `ubuntu-22.04-arm`) uniquely determines a
+        # `runner.arch` (e.g., ARM64), so there is no need to include the latter
+        # as part of the cache key.
+        key: cabal-${{ inputs.os }}-ghc${{ inputs.ghc }}
```

Comment:
It seems like the `inputs.cabal` should be part of this key, otherwise a change to that input will still cache hit and may not actually run anything.

Reply:
Also wondering if this should use actual versions from the the `setup-haskell` step (assuming it exports those) to avoid using "latest" as the key.